Ventcrawling improvements, performance and visual

LemonInTheDark opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Ok so pipecrawl images were updating EVERY TIME YOU MOVED
This was not good mojo, especially because they were iterating ALL the pipes in your net (Ie at least a thousand) each time
![image](https://user-images.githubusercontent.com/58055496/166860865-40a22cc6-df4d-4b08-ae63-29eb855c6b11.png)

Soooo
Adds a concept called a cell tracker datum.
It manages a list of spatial cells a passed in bound is "inside", and when queried will return a list of new cells, and old cells.

Because we only really care about maintaining an absolute window of "CELLS WE ARE IN" and less about always removing cells we're not in, we can manage a second window to prevent moving up and down on a cell line causing a ton of updates.

I've also added atmos machines to the spatial grid. This should have very little overhead outside of memory, because of how rarely they move.

Uses this concept to optimize pipecrawling significantly, from 0.1 deciseconds per call (1/5th OF A TICK) before to roughly 0.07142857ms per call. this is a reduction of over 99%.

Also moves pipecrawl images to their own plane, so they don't overlap ui elements

You can now move in more then one direction when pipecrawling
This works as expected, if you hold up and left, move up for a while, and come to a fork, you'll go left

Added some effects to pipecrawling. It'll darken the lighting plane slightly, so you get a nice effect instead of just fullbright.
Also added a color matrix and drop shadow to the pipe images, this way they stand out a bit more.

You now glide between pipe moves, rather then moving instantly. 
This doesn't effect your actual move rate, but it no longer feels jittery with say, 60fps

Oh also I refactored spatial grid code a bit, partly to add support for this, and partly to make it nicer.

EDIT: spatial grid tracking, minor rework
[Reworks how being interested in the spatial grid is tracked](https://github.com/tgstation/tgstation/pull/66709/commits/34b7b261467383136cc92f65f35f4642e9ef43f9)
[34b7b26](https://github.com/tgstation/tgstation/pull/66709/commits/34b7b261467383136cc92f65f35f4642e9ef43f9)

Rather then checking multiple variables on the atom to consider, we instead check for the existence of a string key.

This key is used by a list on the spatial grid subsystem to retrieve a cached list of all of the atoms "types"

Doing this requires doing a bit of extra work in important_recursive_contents code, but it allows us to separate being a part of the spatial grid from using important recursive contents, which is very nice.

As a consequence, I've had to unroll some lazylist macros in important recursive contents logic. It's not ""that"" bad but it's not great either.

Oh and this adds a slight cost to enter/exit cell, but it's minimal.
Basically, rather then checking for different features of a grid member, we just iterate the list their string key points to. Very handy

So there's an added cost of a list copy and all, but we save the headache of more types technically increasing the cost of addition/removal.

I also made adding/removing from the grid into one "pulbic" proc rather then two different ones for each operation, because it was starting to get silly

## Why It's Good For The Game

Faster, nicer, stronger.

Old

https://user-images.githubusercontent.com/58055496/166861161-bae4e70d-0b1e-46ba-876b-9782f1a98d13.mp4

New

https://user-images.githubusercontent.com/58055496/166861130-ea670191-7a06-4777-9b5c-b5937be58f9c.mp4

Old
![image](https://user-images.githubusercontent.com/58055496/166860865-40a22cc6-df4d-4b08-ae63-29eb855c6b11.png)
New
![image](https://user-images.githubusercontent.com/58055496/166861230-6b1b4efa-e281-4bf0-8faf-b1e33c5aa8df.png)



<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
qol: When ventcrawling, rather then jumping between pipes, your view will "glide"
qol: When ventcrawling, holding two directions no longer locks you in place. Instead behaving as expected
qol: Ventcrawling now darkens the background while you're doing it. Hopefully this looks slightly better then the fullbright of current.
code: Ventcrawling code has had a 99% performance increase. I am not joking. Won't lag the server anymore.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

LemonInTheDark committed:
Initial pipecrawl work

Ok so pipecrawl images were updating EVERY TIME YOU MOVED
This was not good mojo

What I've done here is twofold
First, I ensured pipecrawl updates only when the net changes. This
breaks the current implementation, but I intend on fixing that

Second, I moved our method of getting pipes to the spatial grid
This isn't that great at the moment, but I intend on adding support for
tracking entered/exited cells, which should make this much better

LemonInTheDark committed:
Much faster pipecrawling processing, niceties

Adds a concept called a cell tracker datum.
It manages a list of cells a passed in bound is "inside", and when
queried will return a list of new cells, and old cells.

Because we only really care about maintaining an absolute window of
"CELLS WE ARE IN" and less about always removing cells we're not in, we
can manage a second window to prevent moving up and down on a cell line
causing a ton of updates.

Uses this concept to optimize pipecrawling significantly, from 3ms per
call before to roughly 0.03ms per call.

Also moves pipecrawl images to their own plane, so they don't overlap ui
elements

LemonInTheDark committed:
Pipecrawling effects niceties, direction help

You can now move in more then one direction when pipecrawling
This works as expected, if you hold up and left, move up for a while,
and come to a fork, you'll go left

Added some effects to pipecrawling. It'll darken the lighting plane
slightly, so you get a nice effect instead of just fullbright.
Also added a color matrix and drop shadow to the pipe images, this way
they stand out a bit more.

You now glide between pipe moves, rather then moving instantly. This
doesn't effect your actual move rate, but it no longer feels jittery
with say, 60fps

LemonInTheDark commented:
Oh also, we could in theory give pipecrawling a delay, I just didn't bother to

Kylerace commented on code/__DEFINES/spatial_gridmap.dm:
```diff
@@ -11,6 +12,11 @@
+#define HAS_SPATIAL_GRID_CONTENTS(movable) ((\
+	movable.important_recursive_contents && (\
+	movable.important_recursive_contents[RECURSIVE_CONTENTS_HEARING_SENSITIVE] || movable.important_recursive_contents[RECURSIVE_CONTENTS_CLIENT_MOBS]))\
+	|| movable.flags_1 & SPATIAL_GRID_MANAGED_1)
```
i dont like this. itd be best for code quality reasons if spatial grid contents were made into an element hooked onto the moved signal but >adding _SendSignal() overhead to movement just to do no work more than 16/17ths of the time

ideally there would be one check for whether something is in the grid. question is whats the best way to do that. obviously forcing everything through important_recursive_contents isnt ideal, but the things that are there already on master need to continue working inside nested locs. perhaps the best way to check for this is a trait that is propagated along with important_recursive_contents for the contents type that do that. ill have to think about this

LemonInTheDark replied:
It could totally be a trait, the memory cost would be a thing, but it wouldn't be "that" bad.
If you wanted to really hack it you could implement trait like behavior manually, with some other single var way of indicating a thing being inherently interested in being managed by the grid

LemonInTheDark replied:
34b7b261467383136cc92f65f35f4642e9ef43f9

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -317,10 +328,14 @@ SUBSYSTEM_DEF(spatial_grid)
 
 ///get all grid cells intersecting the bounding box around center with sides of length 2 * range
 /datum/controller/subsystem/spatial_grid/proc/get_cells_in_range(atom/center, range)
+	return get_cells_in_bound(center, range, range)
```
muh overhead

LemonInTheDark replied:
ahhhhhhhhhhhhhhh my never used debug call is 0.000001% more expensive ahhhhhhhhhhh

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -317,10 +328,14 @@ SUBSYSTEM_DEF(spatial_grid)
+	return get_cells_in_bound(center, range, range)
+
+///get all grid cells intersecting the bounding box around center with sides of length (2 * range_x, 2 * range_y)
+/datum/controller/subsystem/spatial_grid/proc/get_cells_in_bound(atom/center, range_x, range_y)
```
bounds

LemonInTheDark replied:
fine fine

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -354,27 +369,32 @@ SUBSYSTEM_DEF(spatial_grid)
+			SEND_SIGNAL(intersecting_cell, SPATIAL_GRID_CELL_ENTERED(RECURSIVE_CONTENTS_HEARING_SENSITIVE), new_target)
 
-		SEND_SIGNAL(intersecting_cell, SPATIAL_GRID_CELL_ENTERED(RECURSIVE_CONTENTS_HEARING_SENSITIVE), new_target)
+	if(istype(new_target, /obj/machinery/atmospherics))
```
i really dont like this. its O(1) now at least but still. this makes me lean towards wanting "x is in the grid channels a, b, and c" to be turned into a trait with several sources but oh god list memory costs

LemonInTheDark replied:
You could use a list for this, maybe a hash -> list global list so it's cheap to store.
Sources gets harder, would need a good reason to want that.

LemonInTheDark replied:
34b7b261467383136cc92f65f35f4642e9ef43f9

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -576,24 +640,36 @@ SUBSYSTEM_DEF(spatial_grid)
 				continue
 			total_client_distance += get_dist(client, other_client)
 
+	for(var/atmos in atmos_list)//n^2 btw
```
n^2 btw

LemonInTheDark replied:
so true up top

Kylerace commented on code/modules/mob/living/living_defines.dm:
```diff
@@ -100,7 +100,12 @@
+	var/last_played_vent = 0
+	/// The last direction we moved in a vent. Used to make holding two directions feel nice
+	var/last_vent_dir = 0
+	/// Cell tracker datum we use to manage the pipes around us, for faster ventcrawling
```
```suggestion
	/// Cell tracker datum we use to manage the pipes around us, for faster ventcrawling.
```

LemonInTheDark replied:
Do I have to

Kylerace replied:
![image](https://user-images.githubusercontent.com/15794172/167100794-c78dbd60-cfef-46db-953c-6d7d0af2b8fa.png)

LemonInTheDark replied:
I don't like using punctuation in my runon sentences alright?

Colovorat commented:
maybe decrease how much it darkens? just a little bit? i think it's too dark

Kylerace commented:
is it just me or is everything in this codebase thats both kinda neglected and a performance problems capable of being optimized to 1/10th the cost

Colovorat commented:
> is it just me or is everything in this codebase thats both kinda neglected and a performance problems capable of being optimized to 1/10th the cost

it's not even 1/10th, it's full on 1/100

Colovorat commented:
> is it just me or is everything in this codebase thats both kinda neglected and a performance problems capable of being optimized to 1/10th the cost

also optimize atmos :trollface:

LemonInTheDark commented on code/__DEFINES/spatial_gridmap.dm:
```diff
@@ -1,7 +1,8 @@
-
-#define SPATIAL_GRID_CELLS_PER_SIDE(world_bounds) ROUND_UP((world_bounds) / SPATIAL_GRID_CELLSIZE)
+///Takes a coordinate, and spits out the spatial grid index (x or y) it's inside
+#define GET_SPATIAL_INDEX(chord) ROUND_UP((chord) / SPATIAL_GRID_CELLSIZE)
```
cohordanite

LemonInTheDark replied:
how do you abbreviate it properly I genuinely don't know

newfren replied:
coord

LemonInTheDark committed:
Bounds

LemonInTheDark commented:
> maybe decrease how much it darkens? just a little bit? i think it's too dark

it admittedly was just kinda a spitball. I don't want to make it too bright tho, I really dislike that aspect of current ventcrawling

LemonInTheDark commented:
> is it just me or is everything in this codebase thats both kinda neglected and a performance problems capable of being optimized to 1/10th the cost

It's not even really optimization, a lot of our code does really stupid shit. If it just wasn't made bad on purpose we'd almost not even need to miro-optimize.

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into pipecrawl

LemonInTheDark committed:
Fixes runtimes, cache something somethign sonic speed

LemonInTheDark commented:
Wouldn't you like to know

LemonInTheDark committed:
Reworks how being interested in the spatial grid is tracked

Rather then checking multiple variables on the atom to consider, we
instead check for the existence of a string key.

This key is used by a list on the spatial grid subsystem to retrive a
cached list of all of the atoms "types"

Doing this requires doing a bit of extra work in
important_recursive_contents code, but it allows us to separate being a
part of the spatial grid from using important recursive contents, which
is very nice.

As a consequence, I've had to unroll some lazylist macros in important
recursive contents logic. It's not ""that"" bad but it's not great
either.

Oh and this adds a slight cost to enter/exit cell, but it's minimal.
Basically, rather then checking for different features of a grid member,
we just iterate the list their string key points to. Very handy

So there's an added cost of a list copy and all, but we save the
headache of more types technically increasing the cost of
addition/removal.

I also made adding/removing from the grid into one "pulbic" proc rather then two
different ones for each operation, because it was starting to get silly

LemonInTheDark committed:
waaa waa it doesn't compile

LemonInTheDark committed:
chord -> coord

LemonInTheDark committed:
Ensures important_recursive_contents is actually emptied on removal

Mothblocks commented on code/__HELPERS/_lists.dm:
```diff
@@ -134,6 +136,7 @@
 	} while(FALSE)
 
 #define SORT_FIRST_INDEX(list) (list[1])
+#define SORT_COMPARE_DIRECTLY(thing) (thing)
```
Unused?

LemonInTheDark replied:
Nope https://github.com/tgstation/tgstation/pull/66709/files#diff-e53779f3667517f78b4cea58163350fc8cd21868ee0977e27d580690c35fd456R525

Mothblocks replied:
The diff you linked me does not appear to point to anything

LemonInTheDark replied:
![image](https://user-images.githubusercontent.com/58055496/167090268-91596839-7461-444f-980d-f49fdd71b142.png)
Fucking github

MrMelbert commented on code/game/atoms_movable.dm:
```diff
@@ -752,36 +756,54 @@
+	for(var/channel in arrived.important_recursive_contents)
+		for(var/atom/movable/location as anything in nested_locs)
+			LAZYINITLIST(location.important_recursive_contents)
+			var/list/recursive_contents = location.important_recursive_contents // blue hedgehod velocity
```
hedgehod

LemonInTheDark replied:
~~soul~~

LemonInTheDark committed:
Removes soul

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -65,6 +68,9 @@
  * [RECURSIVE_CONTENTS_HEARING_SENSITIVE] and [RECURSIVE_CONTENTS_CLIENT_MOBS] because both are those are both 1. important and 2. commonly searched for
+ *
+ * the system addtionally supports pipes, via a direct typecheck
+ * I wouldn't noramally do this, but the case where we want to print pipes is expensive enough I feel I need to
```
this isnt true anymore

Kylerace replied:
the typecheck part i mean

LemonInTheDark replied:
V true

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -358,56 +411,66 @@ SUBSYSTEM_DEF(spatial_grid)
-			GRID_CELL_REMOVE(intersecting_cell.hearing_contents, old_target.important_recursive_contents[SPATIAL_GRID_CONTENTS_TYPE_HEARING])
+			var/list/old_target_contents = old_target.important_recursive_contents
+			GRID_CELL_REMOVE(intersecting_cell.hearing_contents, old_target_contents[SPATIAL_GRID_CONTENTS_TYPE_HEARING])
+			SEND_SIGNAL(intersecting_cell, SPATIAL_GRID_CELL_EXITED(exclusive_type), old_target.important_recursive_contents[SPATIAL_GRID_CONTENTS_TYPE_HEARING])
```
you dont use the old_target_contents list for this signal and the one above

LemonInTheDark replied:
thanks boss

LemonInTheDark replied:
One above?

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -441,13 +505,65 @@ SUBSYSTEM_DEF(spatial_grid)
+
+/// Alerts the atom's current cell that it wishes to be treated as a member
+/// This functionally amounts to "hey, I was recently made aware by [add_grid_awareness], please insert me into my current cell"
+/datum/controller/subsystem/spatial_grid/proc/add_grid_membership(atom/movable/add_to, turf/target_turf, type)
```
could you group the add's with remove's instead of add's with add's and remove's with remove's.

Kylerace replied:
these procs should also be higher up in the file, id say either just above or just below enter_cell() and exit_cell() as a pair

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -74,7 +80,10 @@ SUBSYSTEM_DEF(spatial_grid)
 	///everything that spawns before us is added to this list until we initialize
-	var/list/waiting_to_add_by_type = list(RECURSIVE_CONTENTS_HEARING_SENSITIVE = list(), RECURSIVE_CONTENTS_CLIENT_MOBS = list())
+	var/list/waiting_to_add_by_type = list(SPATIAL_GRID_CONTENTS_TYPE_HEARING = list(), SPATIAL_GRID_CONTENTS_TYPE_CLIENTS = list(), SPATIAL_GRID_CONTENTS_TYPE_ATMOS = list())
+	///associated list of unique string keys -> lists of spatial grid types
+	///we use this to make adding to a cell static cost, and to save on memory
```
```suggestion
	///associative list of the form: movable.spatial_grid_key (string) -> inner list of spatial grid types for that key. 
	///inner lists contain contents channel types such as SPATIAL_GRID_CONTENTS_TYPE_HEARING etc.
	///we use this to make adding to a cell static cost, and to save on memory
```

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -466,30 +582,43 @@ SUBSYSTEM_DEF(spatial_grid)
 /atom/proc/find_grid_statistics_for_z_level(insert_clients = 0)
```
lol this proc does nothing but make it more annoying to add new contents

Kylerace replied:
should probably be removed or redone.

LemonInTheDark replied:
I mean it's at least at barebones useful for checking if everything's working right.
You want me to nuke it?

Kylerace replied:
nah it can stay, i think it can work for every new contents type without needing to be added to if i store contents vars

Kylerace commented on code/game/atoms_movable.dm:
```diff
@@ -48,6 +48,10 @@
 	///only the last container of a client eye has this list assuming no movement since SSparallax's last fire
 	var/list/client_mobs_in_contents
 
+	/// String representing the spatial grid groups we want to be held in
+	/// Key'd to a global list of lists. We do it like this to prevent people trying to mutate them
```
```suggestion
	/// String representing the spatial grid groups we want to be held in.
	/// acts as a key to the list of spatial grid contents types we exist in via SSspatial_grid.spatial_grid_categories.
	/// We do it like this to prevent people trying to mutate them and to save memory on holding the lists ourselves
```

LemonInTheDark committed:
Kyler's review

Co-authored-by: Kylerace <kylerlumpkin1@gmail.com>

LemonInTheDark committed:
Kyler's review 2

Co-authored-by: Kylerace <kylerlumpkin1@gmail.com>

LemonInTheDark committed:
Kyler's review 3

Moves some procs around, improves some documentation, catches a few
small issues

LemonInTheDark committed:
Merge branch 'pipecrawl' of https://github.com/LemonInTheDark/tgstation into pipecrawl
